pkg/agent/util/net_linux.go

```diff
@@ -376,3 +378,151 @@ func renameHostInterface(oriName string, newName string) error {
+		return true, nil
+	}); err != nil {
+		return "", false, fmt.Errorf("failed waiting for internal port to show up: %v", err)
+	}
+
+	localLink, err := netlink.LinkByName(ifaceName)
+	if err != nil {
+		return "", false, err
+	}
+	if _, _, err = SetLinkUp(ifaceName); err != nil {
+		return "", false, fmt.Errorf("failed to set link up: %v", err)
+	}
+
+	// Check if interface is configured with an IPv6 address: if it is, we need to ensure that IPv6
+	// is enabled on the OVS internal port as we need to move all IP addresses over.
+	for _, ip := range ifaceIPs {
+		if ip.IP.To4() == nil {
+			klog.InfoS("Interface has IPv6 address, ensuring that IPv6 is enabled on bridge local port", "port", ifaceName)
+			if err := EnsureIPv6EnabledOnInterface(ifaceName); err != nil {
+				klog.ErrorS(err, "Failed to ensure that IPv6 is enabled on bridge local port, moving uplink IPs to bridge is likely to fail", "port", ifaceName)
+			}
+			break
+		}
+	}
+
+	if err = ConfigureLinkAddresses(localLink.Attrs().Index, ifaceIPs); err != nil {
+		return "", false, err
+	}
+	if err = ConfigureLinkAddresses(iface.Index, nil); err != nil {
+		return "", false, err
+	}
+	// Restore the host routes which are lost when moving the network configuration of the
+	// host interface to OVS bridge interface.
+	if err = ConfigureLinkRoutes(localLink, ifaceRoutes); err != nil {
+		return "", false, err
+	}
+	return bridgedName, false, nil
+}
+
+// RestoreHostInterfaceConfiguration restore the configuration from bridge back to host interface, reverting the
+// actions taken in PrepareHostInterfaceConnection.
+func RestoreHostInterfaceConfiguration(brName string, interfaceName string) {
+	klog.V(4).InfoS("Restoring bridge config to host interface")
+	bridgedName := GenerateUplinkInterfaceName(interfaceName)
+	// restore if interface eth0~ exists
+	if exists, err := interfaceExists(bridgedName); err != nil {
+		klog.ErrorS(err, "Failed to check if interface exists", "interface", bridgedName)
+		return
+	} else if exists {
+		// get interface config
+		var interfaceIPs []*net.IPNet
+		var interfaceRoutes []interface{}
+		if exists, err = interfaceExists(interfaceName); err != nil {
+			klog.ErrorS(err, "Failed to check if interface exists", "interface", interfaceName)
+		} else if exists {
+			_, interfaceIPs, interfaceRoutes, err = GetInterfaceConfig(interfaceName)
+			if err != nil {
+				klog.ErrorS(err, "Failed to get interface config", "interface", interfaceName)
+			}
+		}
+		// remove host interface (eth0~) from bridge
+		if err = deleteOVSPort(brName, bridgedName); err != nil {
+			klog.ErrorS(err, "Delete OVS port failed", "port", bridgedName)
+		}
+
+		// delete internal port (eth0)
+		if err = deleteOVSPort(brName, interfaceName); err != nil {
+			klog.ErrorS(err, "Delete OVS port failed", "port", bridgedName)
+		}
+
+		// rename host interface(eth0~ -> eth0)
+		if err = RenameInterface(bridgedName, interfaceName); err != nil {
```

Comment:
Add earlier comments back:

From Jianjun:
> But here, even interfaceName does not exist, we still want to rename eth0~ back to eth0.
> Maybe we can add a boolean to record if an interface exists or not. Maybe the code is simpler to return at errors, or when bridgedName does not exist.

From Daman:
> This part will only be executed when both "eth0" and "eth0~" exist at the beginning of restoration.
> We don't want the rename operation when "eth0~" doesn't exist right?

Reply:
I meant it is possible "eth0~" exists but "eth0" does not, when interface renaming succeeded but bridge port creation failed. In this case, we should still change "eth0~" back to "eth0".

Reply:
With the latest push, I'm renaming the interface even if **eth0** doesn't exist.
We skip restoration completely if **eth0~** doesn't exist. In case **eth0** doesn't exist we only skip get interface config.

Reply:
Ok. Sorry I missed that.